
Every kernel release there are a few changes to headers
made. Some code gets shifted around between headers or
new headers are defined. This patch deals with such
cases.

This file also deals with enhancements with new kernel headers
which older kernels would conflict with.

--- a/drivers/net/wireless/b43/phy_common.h
+++ b/drivers/net/wireless/b43/phy_common.h
@@ -2,6 +2,9 @@
 #define LINUX_B43_PHY_COMMON_H_
 
 #include <linux/types.h>
+#if (LINUX_VERSION_CODE == KERNEL_VERSION(2,6,28))
+#include <asm/atomic.h>
+#endif
 
 struct b43_wldev;
 
--- a/drivers/net/wireless/wl12xx/wl1251_main.c	2009-08-07 11:42:12.532495607 -0700
+++ b/drivers/net/wireless/wl12xx/wl1251_main.c	2009-08-07 12:08:31.167548696 -0700
@@ -26,6 +26,9 @@
 #include <linux/firmware.h>
 #include <linux/delay.h>
 #include <linux/irq.h>
+#if (LINUX_VERSION_CODE == KERNEL_VERSION(2,6,28))
+#include <linux/device.h>
+#endif
 #include <linux/crc32.h>
 #include <linux/etherdevice.h>
 
--- a/drivers/net/wireless/wl12xx/wl1251_spi.c	2009-08-07 12:24:58.907526225 -0700
+++ b/drivers/net/wireless/wl12xx/wl1251_spi.c	2009-08-07 12:24:59.199494403 -0700
@@ -24,6 +24,9 @@
 #include <linux/irq.h>
 #include <linux/module.h>
 #include <linux/crc7.h>
+#if (LINUX_VERSION_CODE == KERNEL_VERSION(2,6,28))
+#include <linux/device.h>
+#endif
 #include <linux/spi/spi.h>
 #include <linux/spi/wl12xx.h>
 
--- a/net/mac80211/driver-trace.h	2009-08-18 16:19:34.574352433 -0700
+++ b/net/mac80211/driver-trace.h	2009-08-18 16:19:35.694367614 -0700
@@ -1,7 +1,9 @@
 #if !defined(__MAC80211_DRIVER_TRACE) || defined(TRACE_HEADER_MULTI_READ)
 #define __MAC80211_DRIVER_TRACE
 
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,27))
 #include <linux/tracepoint.h>
+#endif /* (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,27)) */
 #include <net/mac80211.h>
 #include "ieee80211_i.h"
 
@@ -669,4 +671,6 @@ TRACE_EVENT(drv_ampdu_action,
 #define TRACE_INCLUDE_PATH .
 #undef TRACE_INCLUDE_FILE
 #define TRACE_INCLUDE_FILE driver-trace
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,30))
 #include <trace/define_trace.h>
+#endif /* (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,30)) */
--- a/net/mac80211/key.c
+++ b/net/mac80211/key.c
@@ -21,6 +21,9 @@
 #include "aes_ccm.h"
 #include "aes_cmac.h"
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,29))
+#include <asm/unaligned.h>
+#endif
 
 /**
  * DOC: Key handling basics
--- a/include/linux/bitops.h	2010-01-05 21:18:40.000000000 -0800
+++ b/include/linux/bitops.h	2010-01-05 21:19:15.000000000 -0800
@@ -7,8 +7,10 @@
 #define BIT_MASK(nr)		(1UL << ((nr) % BITS_PER_LONG))
 #define BIT_WORD(nr)		((nr) / BITS_PER_LONG)
 #define BITS_PER_BYTE		8
+#ifndef BITS_TO_LONGS
 #define BITS_TO_LONGS(nr)	DIV_ROUND_UP(nr, BITS_PER_BYTE * sizeof(long))
 #endif
+#endif
 
 /*
  * Include this here because some architectures need generic_ffs/fls in
--- a/net/mac80211/main.c	2010-01-05 21:54:51.000000000 -0800
+++ b/net/mac80211/main.c	2010-01-05 21:55:21.000000000 -0800
@@ -22,7 +22,9 @@
 #include <linux/rtnetlink.h>
 #include <linux/bitmap.h>
 #include <linux/pm_qos_params.h>
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,24))
 #include <net/net_namespace.h>
+#endif
 #include <net/cfg80211.h>
 
 #include "ieee80211_i.h"
--- a/net/mac80211/cfg.c	2010-01-05 21:59:44.000000000 -0800
+++ b/net/mac80211/cfg.c	2010-01-05 22:00:16.000000000 -0800
@@ -9,7 +9,9 @@
 #include <linux/ieee80211.h>
 #include <linux/nl80211.h>
 #include <linux/rtnetlink.h>
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,24))
 #include <net/net_namespace.h>
+#endif
 #include <linux/rcupdate.h>
 #include <net/cfg80211.h>
 #include "ieee80211_i.h"
--- a/net/mac80211/tx.c	2010-01-05 22:01:11.000000000 -0800
+++ b/net/mac80211/tx.c	2010-01-05 22:01:28.000000000 -0800
@@ -18,7 +18,9 @@
 #include <linux/etherdevice.h>
 #include <linux/bitmap.h>
 #include <linux/rcupdate.h>
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,24))
 #include <net/net_namespace.h>
+#endif
 #include <net/ieee80211_radiotap.h>
 #include <net/cfg80211.h>
 #include <net/mac80211.h>
--- a/net/mac80211/util.c	2010-01-05 22:02:59.000000000 -0800
+++ b/net/mac80211/util.c	2010-01-05 22:03:30.000000000 -0800
@@ -21,7 +21,9 @@
 #include <linux/wireless.h>
 #include <linux/bitmap.h>
 #include <linux/crc32.h>
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,24))
 #include <net/net_namespace.h>
+#endif
 #include <net/cfg80211.h>
 #include <net/rtnetlink.h>
 
--- a/include/linux/bitops.h	2010-01-05 22:22:52.000000000 -0800
+++ b/include/linux/bitops.h	2010-01-05 22:32:25.000000000 -0800
@@ -3,7 +3,9 @@
 #include <asm/types.h>
 
 #ifdef	__KERNEL__
+#ifndef BIT
 #define BIT(nr)			(1UL << (nr))
+#endif
 #define BIT_MASK(nr)		(1UL << ((nr) % BITS_PER_LONG))
 #define BIT_WORD(nr)		((nr) / BITS_PER_LONG)
 #define BITS_PER_BYTE		8
